File: token.cpp
/**
 * Copyright (c) 2022 Atsushi Komaba
 *
 * This file is part of Syscraws.
 * Syscraws is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or any later version.
 * Syscraws is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
 * You should have received a copy of the GNU General Public License along with Syscraws. If not, see <https://www.gnu.org/licenses/>. 
 *
 * @file token.cpp
 */
#include "token.hpp"
#include "error.hpp"

namespace token {
    Token::~Token() = default;
    Identifier::Identifier(std::string_view name): name(name) {}
    Number::Number(std::string_view value): value(value) {}
    String::String(std::string value) : value(std::move(value)) {}

    std::unique_ptr<pre_ast::Term> Token::factor(){ return nullptr; }
    std::unique_ptr<pre_ast::Term> Identifier::factor(){ return std::make_unique<pre_ast::Identifier>(name); }
    std::unique_ptr<pre_ast::Term> Number::factor(){
        if(value[0] == '0' && value.length() >= 2){
            if(value[1] == 'b'){
                bool has_digit = false;
                for(std::size_t i = 2; i < value.length(); i++){
                    if(value[i] == '_') continue;
                    else if(value[i] == '0' || value[i] == '1');
                    else throw error::make<error::InvalidNumericLiteral>(std::move(pos));
                    has_digit = true;
                }
                if(has_digit) return std::make_unique<pre_ast::BinInt>(value.substr(2));
                else throw error::make<error::InvalidNumericLiteral>(std::move(pos));
            }else if(value[1] == 'o'){
                bool has_digit = false;
                for(std::size_t i = 2; i < value.length(); i++){
                    if(value[i] == '_') continue;
                    if('0' <= value[i] && value[i] <= '8');
                    else throw error::make<error::InvalidNumericLiteral>(std::move(pos));
                    has_digit = true;
                }
                if(has_digit) return std::make_unique<pre_ast::OctInt>(value.substr(2));
                else throw error::make<error::InvalidNumericLiteral>(std::move(pos));
            }else if(value[1] == 'x'){
                bool has_digit = false;
                for(std::size_t i = 2; i < value.length(); i++){
                    if(value[i] == '_') continue;
                    else if('0' <= value[i] && value[i] <= '9');
                    else if('a' <= value[i] && value[i] <= 'f');
                    else if('A' <= value[i] && value[i] <= 'F');
                    else throw error::make<error::InvalidNumericLiteral>(std::move(pos));
                    has_digit = true;
                }
                if(has_digit) return std::make_unique<pre_ast::HexInt>(value.substr(2));
                else throw error::make<error::InvalidNumericLiteral>(std::move(pos));
            }
        }
        for(std::size_t i = 0; i < value.length(); i++){
            if(('0' <= value[i] && value[i] <= '9') || value[i] == '_'){
            }else{
                return std::make_unique<pre_ast::Float>(value);
            }
        }
        return std::make_unique<pre_ast::DecInt>(value);
    }
    std::unique_ptr<pre_ast::Term> String::factor(){ return std::make_unique<pre_ast::String>(std::move(value)); }

    std::optional<Keyword> Token::keyword(){ return std::nullopt; }
    std::optional<Keyword> Identifier::keyword(){
        if(name == "if") return Keyword::If;
        else if(name == "else") return Keyword::Else;
        else if(name == "while") return Keyword::While;
        else if(name == "break") return Keyword::Break;
        else if(name == "continue") return Keyword::Continue;
        else if(name == "return") return Keyword::Return;
        else return std::nullopt;
    }

    std::optional<pre_ast::UnaryOperator> Token::prefix(){ return std::nullopt; }
    std::optional<pre_ast::UnaryOperator> Plus::prefix(){ return pre_ast::UnaryOperator::Plus; }
    std::optional<pre_ast::UnaryOperator> Hyphen::prefix(){ return pre_ast::UnaryOperator::Minus; }
    std::optional<pre_ast::UnaryOperator> DoublePlus::prefix(){ return pre_ast::UnaryOperator::PreInc; }
    std::optional<pre_ast::UnaryOperator> DoubleHyphen::prefix(){ return pre_ast::UnaryOperator::PreDec; }
    std::optional<pre_ast::UnaryOperator> Slash::prefix(){ return pre_ast::UnaryOperator::Recip; }
    std::optional<pre_ast::UnaryOperator> Exclamation::prefix(){ return pre_ast::UnaryOperator::LogicalNot; }
    std::optional<pre_ast::UnaryOperator> Tilde::prefix(){ return pre_ast::UnaryOperator::BitNot; }

    std::optional<pre_ast::UnaryOperator> Token::suffix(){ return std::nullopt; }
    std::optional<pre_ast::UnaryOperator> DoublePlus::suffix(){ return pre_ast::UnaryOperator::PostInc; }
    std::optional<pre_ast::UnaryOperator> DoubleHyphen::suffix(){ return pre_ast::UnaryOperator::PostDec; }

    std::optional<pre_ast::BinaryOperator> Token::infix(){ return std::nullopt; }
    std::optional<pre_ast::BinaryOperator> Plus::infix(){ return pre_ast::BinaryOperator::Add; }
    std::optional<pre_ast::BinaryOperator> Hyphen::infix(){ return pre_ast::BinaryOperator::Sub; }
    std::optional<pre_ast::BinaryOperator> Asterisk::infix(){ return pre_ast::BinaryOperator::Mul; }
    std::optional<pre_ast::BinaryOperator> Slash::infix(){ return pre_ast::BinaryOperator::Div; }
    std::optional<pre_ast::BinaryOperator> Percent::infix(){ return pre_ast::BinaryOperator::Rem; }
    std::optional<pre_ast::BinaryOperator> DoubleGreater::infix(){ return pre_ast::BinaryOperator::RightShift; }
    std::optional<pre_ast::BinaryOperator> DoubleLess::infix(){ return pre_ast::BinaryOperator::LeftShift; }
    std::optional<pre_ast::BinaryOperator> TripleGreater::infix(){ return pre_ast::BinaryOperator::ForwardShift; }
    std::optional<pre_ast::BinaryOperator> TripleLess::infix(){ return pre_ast::BinaryOperator::BackwardShift; }
    std::optional<pre_ast::BinaryOperator> DoubleEqual::infix(){ return pre_ast::BinaryOperator::Equal; }
    std::optional<pre_ast::BinaryOperator> ExclamationEqual::infix(){ return pre_ast::BinaryOperator::NotEqual; }
    std::optional<pre_ast::BinaryOperator> Less::infix(){ return pre_ast::BinaryOperator::Less; }
    std::optional<pre_ast::BinaryOperator> LessEqual::infix(){ return pre_ast::BinaryOperator::LessEqual; }
    std::optional<pre_ast::BinaryOperator> Greater::infix(){ return pre_ast::BinaryOperator::Greater; }
    std::optional<pre_ast::BinaryOperator> GreaterEqual::infix(){ return pre_ast::BinaryOperator::GreaterEqual; }
    std::optional<pre_ast::BinaryOperator> DoubleAmpersand::infix(){ return pre_ast::BinaryOperator::LogicalAnd; }
    std::optional<pre_ast::BinaryOperator> DoubleBar::infix(){ return pre_ast::BinaryOperator::LogicalOr; }
    std::optional<pre_ast::BinaryOperator> Ampersand::infix(){ return pre_ast::BinaryOperator::BitAnd; }
    std::optional<pre_ast::BinaryOperator> Bar::infix(){ return pre_ast::BinaryOperator::BitOr; }
    std::optional<pre_ast::BinaryOperator> Circumflex::infix(){ return pre_ast::BinaryOperator::BitXor; }
    std::optional<pre_ast::BinaryOperator> Colon::infix(){ return pre_ast::BinaryOperator::Type; }
    std::optional<pre_ast::BinaryOperator> Equal::infix(){ return pre_ast::BinaryOperator::Assign; }
    std::optional<pre_ast::BinaryOperator> PlusEqual::infix(){ return pre_ast::BinaryOperator::AddAssign; }
    std::optional<pre_ast::BinaryOperator> HyphenEqual::infix(){ return pre_ast::BinaryOperator::SubAssign; }
    std::optional<pre_ast::BinaryOperator> AsteriskEqual::infix(){ return pre_ast::BinaryOperator::MulAssign; }
    std::optional<pre_ast::BinaryOperator> SlashEqual::infix(){ return pre_ast::BinaryOperator::DivAssign; }
    std::optional<pre_ast::BinaryOperator> PercentEqual::infix(){ return pre_ast::BinaryOperator::RemAssign; }
    std::optional<pre_ast::BinaryOperator> AmpersandEqual::infix(){ return pre_ast::BinaryOperator::BitAndAssign; }
    std::optional<pre_ast::BinaryOperator> BarEqual::infix(){ return pre_ast::BinaryOperator::BitOrAssign; }
    std::optional<pre_ast::BinaryOperator> CircumflexEqual::infix(){ return pre_ast::BinaryOperator::BitXorAssign; }
    std::optional<pre_ast::BinaryOperator> DoubleGreaterEqual::infix(){ return pre_ast::BinaryOperator::RightShiftAssign; }
    std::optional<pre_ast::BinaryOperator> DoubleLessEqual::infix(){ return pre_ast::BinaryOperator::LeftShiftAssign; }
    std::optional<pre_ast::BinaryOperator> TripleGreaterEqual::infix(){ return pre_ast::BinaryOperator::ForwardShiftAssign; }
    std::optional<pre_ast::BinaryOperator> TripleLessEqual::infix(){ return pre_ast::BinaryOperator::BackwardShiftAssign; }

    bool Token::is_comma() const { return false; }
    bool Comma::is_comma() const { return true; }
    bool Token::is_semicolon() const { return false; }
    bool Semicolon::is_semicolon() const { return true; }
    bool Token::is_opening_parenthesis() const { return false; }
    bool OpeningParenthesis::is_opening_parenthesis() const { return true; }
    bool Token::is_closing_parenthesis() const { return false; }
    bool ClosingParenthesis::is_closing_parenthesis() const { return true; }
    bool Token::is_opening_bracket() const { return false; }
    bool OpeningBracket::is_opening_bracket() const { return true; }
    bool Token::is_closing_bracket() const { return false; }
    bool ClosingBracket::is_closing_bracket() const { return true; }
    bool Token::is_opening_brace() const { return false; }
    bool OpeningBrace::is_opening_brace() const { return true; }
    bool Token::is_closing_brace() const { return false; }
    bool ClosingBrace::is_closing_brace() const { return true; }

    std::optional<pre_ast::BracketType> Token::opening_bracket_type() const { return std::nullopt; }
    std::optional<pre_ast::BracketType> Token::closing_bracket_type() const { return std::nullopt; }
    std::optional<pre_ast::BracketType> OpeningParenthesis::opening_bracket_type() const { return pre_ast::BracketType::Round; }
    std::optional<pre_ast::BracketType> ClosingParenthesis::closing_bracket_type() const { return pre_ast::BracketType::Round; }
    std::optional<pre_ast::BracketType> OpeningBracket::opening_bracket_type() const { return pre_ast::BracketType::Square; }
    std::optional<pre_ast::BracketType> ClosingBracket::closing_bracket_type() const { return pre_ast::BracketType::Square; }
}

#ifdef DEBUG
#include <iostream>
static void indent(int depth){
    for(int i = 0; i < depth; i++) std::cout << "    ";
}
namespace token {
    void Identifier::debug_print(int depth) const {
        indent(depth);
        std::cout << "identifier (" << name << ")" << std::endl;
    }
    void Number::debug_print(int depth) const {
        indent(depth);
        std::cout << "number (" << value << ")" << std::endl;
    }
    void String::debug_print(int depth) const {
        indent(depth);
        std::cout << "string (" << value << ")" << std::endl;
    }

#define define_debug_print(token, text) \
    void token::debug_print(int depth) const { \
        indent(depth); \
        std::cout << text << std::endl; \
    } \

    define_debug_print(Plus, "plus (+)")
    define_debug_print(DoublePlus, "double plus (++)")
    define_debug_print(PlusEqual, "plus equal (+=)")
    define_debug_print(Hyphen, "hyphen (-)")
    define_debug_print(DoubleHyphen, "double hyphen (--)")
    define_debug_print(HyphenEqual, "hyphen equal (-=)")
    define_debug_print(Asterisk, "asterisk (*)")
    define_debug_print(AsteriskEqual, "asterisk equal (*=)")
    define_debug_print(Slash, "slash (/)")
    define_debug_print(SlashEqual, "slash equal (/=)")
    define_debug_print(Percent, "percent (%)")
    define_debug_print(PercentEqual, "percent equal (%=)")
    define_debug_print(Equal, "equal (=)")
    define_debug_print(DoubleEqual, "double equal (==)")
    define_debug_print(Exclamation, "exclamation (!)")
    define_debug_print(ExclamationEqual, "exclamation equal (!=)")
    define_debug_print(Less, "less (<)")
    define_debug_print(LessEqual, "less equal (<=)")
    define_debug_print(DoubleLess, "double less (<<)")
    define_debug_print(DoubleLessEqual, "double less equal (<<=)")
    define_debug_print(TripleLess, "triple less (<<<)")
    define_debug_print(TripleLessEqual, "triple less equal (<<<=)")
    define_debug_print(Greater, "greater (>)")
    define_debug_print(GreaterEqual, "greater equal (>=)")
    define_debug_print(DoubleGreater, "double greater (>>)")
    define_debug_print(DoubleGreaterEqual, "double greater equal (>>=)")
    define_debug_print(TripleGreater, "triple greater (>>>)")
    define_debug_print(TripleGreaterEqual, "triple greater equal (>>>=)")
    define_debug_print(Ampersand, "ampersand (&)")
    define_debug_print(AmpersandEqual, "ampersand equal (&=)")
    define_debug_print(DoubleAmpersand, "double ampersand (&&)")
    define_debug_print(Bar, "bar (|)")
    define_debug_print(BarEqual, "bar equal (|=)")
    define_debug_print(DoubleBar, "double bar (||)")
    define_debug_print(Circumflex, "circumflex (^)")
    define_debug_print(CircumflexEqual, "circumflex equal (^=)")
    define_debug_print(Dot, "dot (.)")
    define_debug_print(Colon, "colon (:)")
    define_debug_print(Semicolon, "semicolon (;)")
    define_debug_print(Comma, "comma (,)")
    define_debug_print(Question, "question (?)")
    define_debug_print(Hash, "hash (#)")
    define_debug_print(Tilde, "tilde (~)")
    define_debug_print(OpeningParenthesis, "opening parenthesis '('")
    define_debug_print(ClosingParenthesis, "closing parenthesis ')'")
    define_debug_print(OpeningBracket, "opening bracket '['")
    define_debug_print(ClosingBracket, "closing bracket ']'")
    define_debug_print(OpeningBrace, "opening brace '{'")
    define_debug_print(ClosingBrace, "closing brace '}'")
}
#endif

File: src/parser/token.rs
/*
 * Copyright (c) 2023 Atsushi Komaba
 *
 * This file is part of Syscraws.
 * Syscraws is free software: you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation, either version 3
 * of the License, or any later version.
 *
 * Syscraws is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Syscraws. If not, see <https://www.gnu.org/licenses/>.
 */

use super::CharsPeekable;
use super::Error;
use either::Either;

use crate::pre_ast::PTerm;

pub enum Token<'id> {
    Identifier(&'id str),
    KeywordIf,
    KeywordElse,
    KeywordWhile,
    KeywordReturn,
    Number(&'id str),
    String(Vec<Either<String, Option<PTerm<'id>>>>),
    Plus,
    PlusEqual,
    DoublePlus,
    Hyphen,
    HyphenEqual,
    DoubleHyphen,
    HyphenGreater,
    Asterisk,
    AsteriskEqual,
    Slash,
    SlashEqual,
    Percent,
    PercentEqual,
    Equal,
    DoubleEqual,
    EqualGreater,
    Exclamation,
    ExclamationEqual,
    Greater,
    GreaterEqual,
    DoubleGreater,
    DoubleGreaterEqual,
    TripleGreater,
    TripleGreaterEqual,
    Less,
    LessEqual,
    DoubleLess,
    DoubleLessEqual,
    TripleLess,
    TripleLessEqual,
    Ampersand,
    AmpersandEqual,
    DoubleAmpersand,
    Bar,
    BarEqual,
    DoubleBar,
    Circumflex,
    CircumflexEqual,
    Dot,
    Colon,
    Semicolon,
    Comma,
    Question,
    Hash,
    Tilde,
    OpeningParenthesis,
    ClosingParenthesis,
    OpeningBracket,
    ClosingBracket,
    OpeningBrace,
    ClosingBrace,
}

pub fn next<'id>(chars: &mut CharsPeekable<'id>) -> Result<Option<Token<'id>>, Error> {
    chars.consume_while(|ch| ch.is_ascii_whitespace());
    let start = chars.offset();
    let Some(first_ch) = chars.next() else {
        return Ok(None);
    };
    let token = match first_ch {
        'a'..='z' | 'A'..='Z' | '_' => {
            chars.consume_while(|ch| matches!(ch, 'a'..='z' | 'A'..='Z' | '0'..='9' | '_'));
            let end = chars.offset();
            Token::Identifier(unsafe { chars.input.get_unchecked(start..end) })
        }
        '0'..='9' => {
            read_number(chars);
            let end = chars.offset();
            Token::Number(unsafe { chars.input.get_unchecked(start..end) })
        }
        '.' => {
            if chars.consume_if(|ch| ch.is_ascii_digit()) {
                read_number(chars);
                let end = chars.offset();
                Token::Number(unsafe { chars.input.get_unchecked(start..end) })
            } else {
                Token::Dot
            }
        }
        '+' => {
            if chars.consume_if_eq('+') {
                Token::DoublePlus
            } else if chars.consume_if_eq('=') {
                Token::PlusEqual
            } else {
                Token::Plus
            }
        }
        '-' => {
            if chars.consume_if_eq('-') {
                Token::DoubleHyphen
            } else if chars.consume_if_eq('=') {
                Token::HyphenEqual
            } else if chars.consume_if_eq('>') {
                Token::HyphenGreater
            } else {
                Token::Hyphen
            }
        }
        '*' => {
            if chars.consume_if_eq('=') {
                Token::AsteriskEqual
            } else {
                Token::Asterisk
            }
        }
        '/' => {
            if chars.consume_if_eq('/') {
                chars.consume_while(|ch| ch != '\n');
                return next(chars);
            } else if chars.consume_if_eq('*') {
                let mut comments = vec![start];
                while !comments.is_empty() {
                    let pos = chars.offset();
                    match chars.next() {
                        Some('/') if chars.consume_if_eq('*') => comments.push(pos),
                        Some('*') if chars.consume_if_eq('/') => {
                            comments.pop();
                        }
                        Some(_) => {}
                        None => return Err(Error::UnterminatedComment(comments)),
                    }
                }
                return next(chars);
            } else if chars.consume_if_eq('=') {
                Token::SlashEqual
            } else {
                Token::Slash
            }
        }
        '%' => {
            if chars.consume_if_eq('=') {
                Token::PercentEqual
            } else {
                Token::Percent
            }
        }
        '=' => {
            if chars.consume_if_eq('=') {
                Token::DoubleEqual
            } else if chars.consume_if_eq('>') {
                Token::EqualGreater
            } else {
                Token::Equal
            }
        }
        '!' => {
            if chars.consume_if_eq('=') {
                Token::ExclamationEqual
            } else {
                Token::Exclamation
            }
        }
        '>' => {
            if chars.consume_if_eq('>') {
                if chars.consume_if_eq('>') {
                    if chars.consume_if_eq('=') {
                        Token::TripleGreaterEqual
                    } else {
                        Token::TripleGreater
                    }
                } else if chars.consume_if_eq('=') {
                    Token::DoubleGreaterEqual
                } else {
                    Token::DoubleGreater
                }
            } else if chars.consume_if_eq('=') {
                Token::GreaterEqual
            } else {
                Token::Greater
            }
        }
        '<' => {
            if chars.consume_if_eq('<') {
                if chars.consume_if_eq('<') {
                    if chars.consume_if_eq('=') {
                        Token::TripleLessEqual
                    } else {
                        Token::TripleLess
                    }
                } else if chars.consume_if_eq('=') {
                    Token::DoubleLessEqual
                } else {
                    Token::DoubleLess
                }
            } else if chars.consume_if_eq('=') {
                Token::LessEqual
            } else {
                Token::Less
            }
        }
        '&' => {
            if chars.consume_if_eq('&') {
                Token::DoubleAmpersand
            } else if chars.consume_if_eq('=') {
                Token::AmpersandEqual
            } else {
                Token::Ampersand
            }
        }
        '|' => {
            if chars.consume_if_eq('|') {
                Token::DoubleBar
            } else if chars.consume_if_eq('=') {
                Token::BarEqual
            } else {
                Token::Bar
            }
        }
        '^' => {
            if chars.consume_if_eq('=') {
                Token::CircumflexEqual
            } else {
                Token::Circumflex
            }
        }
        ':' => Token::Colon,
        ';' => Token::Semicolon,
        ',' => Token::Comma,
        '?' => Token::Question,
        '#' => Token::Hash,
        '~' => Token::Tilde,
        '(' => Token::OpeningParenthesis,
        ')' => Token::ClosingParenthesis,
        '[' => Token::OpeningBracket,
        ']' => Token::ClosingBracket,
        '{' => Token::OpeningBrace,
        '}' => Token::ClosingBrace,
        '"' => {
            let mut components = Vec::new();
            let mut string = String::new();
            #[derive(PartialEq, Eq)]
            enum Action {
                Border,
                Char,
                Expr,
            }
            let mut prev_action = Action::Border;
            loop {
                let Some(mut ch) = chars.next() else {
                    return Err(Error::UnterminatedStringLiteral(start));
                };
                let action = match ch {
                    '"' => Action::Border,
                    '{' => Action::Expr,
                    '\\' => {
                        ch = match chars.next() {
                            Some('n') => '\n',
                            Some('r') => '\r',
                            Some('t') => '\t',
                            Some('"') => '\"',
                            Some('\\') => '\\',
                            Some('0') => '\0',
                            Some('\'') => '\'',
                            _ => return Err(Error::InvalidEscapeSequence(start)),
                        };
                        Action::Char
                    }
                    _ => Action::Char,
                };
                if action == Action::Char {
                    string.push(ch);
                } else if prev_action == Action::Char {
                    components.push(Either::Left(std::mem::take(&mut string)))
                }
                if action == Action::Expr {
                    let mut peeked = next(chars)?;
                    let term = crate::parser::parse_assign(chars, &mut peeked);
                    assert!(matches!(peeked, Some(Token::ClosingBrace)));
                    components.push(Either::Right(term.unwrap()));
                } else if action == Action::Border {
                    break;
                }
                prev_action = action;
            }
            Token::String(components)
        }
        _ => return Err(Error::UnexpectedCharacter(start)),
    };
    Ok(Some(token))
}

fn read_number(chars: &mut CharsPeekable) {
    let mut after_e = false;
    chars.consume_while(|ch| {
        after_e = match ch {
            'e' | 'E' => true,
            '0'..='9' | 'a'..='z' | 'A'..='Z' | '_' | '.' => false,
            '+' | '-' if after_e => false,
            _ => return false,
        };
        true
    });
}
